src/algebra/ring.lean

```diff
@@ -505,8 +565,105 @@ def mk' {γ} [semiring α] [ring γ] (f : α →* γ) (map_add : ∀ a b : α, f
 
 end ring_hom
 
-section prio
-set_option default_priority 100 -- see Note [default priority]
+@[ancestor ring comm_semigroup]
+class comm_ring (α : Type u) extends ring α, comm_semigroup α
+
+instance comm_ring.to_comm_semiring [s : comm_ring α] : comm_semiring α :=
+{ mul_zero := mul_zero, zero_mul := zero_mul, ..s }
+
+section comm_ring
+variables [comm_ring α] {a b c : α}
+
+local attribute [simp] add_assoc add_comm add_left_comm mul_comm
+
+lemma mul_self_sub_mul_self_eq (a b : α) : a * a - b * b = (a + b) * (a - b) :=
+begin simp [right_distrib, left_distrib, sub_eq_add_neg] end
+
+lemma mul_self_sub_one_eq (a : α) : a * a - 1 = (a + 1) * (a - 1) :=
+begin simp [right_distrib, left_distrib, sub_eq_add_neg], rw [add_left_comm, add_comm (-a), add_left_comm a], simp end
+
+lemma add_mul_self_eq (a b : α) : (a + b) * (a + b) = a*a + 2*a*b + b*b :=
+calc (a + b)*(a + b) = a*a + (1+1)*a*b + b*b : by simp [right_distrib, left_distrib]
+              ...     = a*a + 2*a*b + b*b     : by rw one_add_one_eq_two
```

Comment:
@kckennylau pointed out that this can be moved to the `comm_semiring` section up above: https://github.com/leanprover-community/lean/issues/232

Reply:
I will postpone that for the next PR.